README.md
![Hail.to](docs/images/hail-logo.png "hail.to")

# Elemental module for SilverStripe Hail

Silverstripe 4 module to integrate the Hail module in SilverStripe Elemental 

## Features

* Add a new element to Elemental type: Hail Carousel
* Ready to use with Bootstrap 4.1 styles and templates
* Responsive configuration for the Carousel
* Mobile friendly using touch events to slide and responsive design

## Requirements

* [firebrandhq/silverstripe-hail ^4.0](https://github.com/firebrandhq/silverstripe-hail)
* [dnadesign/silverstripe-elemental ^2.0](https://github.com/dnadesign/silverstripe-elemental)
* [sheadawson/silverstripe-dependentdropdownfield ^2.0](https://github.com/sheadawson/silverstripe-dependentdropdownfield)
* jQuery and Bootstrap 4.1+ (included from the SilverStripe Hail module), [see jQuery and Boostrap requirements](#jquery-and-bootstrap-requirements)

## Installation

**Run the following command:**

```sh
composer require firebrandhq/silverstripe-hail-elemental "^1.0"
```

**Perform a dev/build (from sake or from a browser)**

## jQuery and Bootstrap requirements

We include jQuery 3.3.1 and Bootstrap 4.1 (javascript and css) from our SilverStripe Hail module

If you need to include your own jQuery and/or Bootstrap (If you compiled Bootstrap from source or want to include those globally for example), simply block our requirement(s) by adding one or all the following to your PageController init() function:

```php
protected function init()
{
    parent::init();
    // You can include any CSS or JS required by your project here.
    // See: https://docs.silverstripe.org/en/developer_guides/templates/requirements/
    
    \SilverStripe\View\Requirements::block('firebrand/silverstripe-hail: thirdparty/bootstrap/styles/bootstrap.min.css');
    \SilverStripe\View\Requirements::block('firebrand/silverstripe-hail: thirdparty/jquery/js/jquery.min.js');
    \SilverStripe\View\Requirements::block('firebrand/silverstripe-hail: thirdparty/bootstrap/js/bootstrap.bundle.min.js');
}
```

## Usage

If SilverStripe Elemental is enabled on the page you should see a Hail Carousel option in the list of elements you can add.
The configuration of the Hail Carousel should be self explanatory, see screenshot below. 

You can configure the responsive breakdowns and carousel behavior from the *Style* tab.
Hail Carousel configuration:

![Hail Carousel config 1](docs/images/screen-1.PNG "Hail Carousel Config 1")

![Hail Carousel config 2](docs/images/screen-2.PNG "Hail Carousel Config 2")

Once configured and published your Hail Carousel should look like this (using default styles and templates):

![Hail Carousel](docs/images/screen-3.PNG "Hail Carousel")

## Contributing

[See CONTRIBUTING.md](CONTRIBUTING.md)

## Versioning

This library follows [Semver](http://semver.org). According to Semver, you will be able to upgrade to any minor or patch version of this library without any breaking changes to the public API. Semver also requires that we clearly define the public API for this library.

All methods, with `public` visibility, are part of the public API. All other methods are not part of the public API. Where possible, we'll try to keep `protected` methods backwards-compatible in minor/patch versions, but if you're overriding methods then please test your work before upgrading.

## Reporting Issues

Please [create an issue](https://github.com/firebrandhq/silverstripe-hail-elemental/issues) for any bugs you've found, or features you're missing.  

## Credits

[Slick Carousel by kenwheeler](https://github.com/kenwheeler/slick/)



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. This project adheres to [Semantic Versioning](http://semver.org/).

## [1.x-dev]

* Initial version
* Silverstripe ^4.1 support

webpack.common.js
// Base
const path = require('path');
const autoprefixer = require('autoprefixer');
const cssnano = require('cssnano');

// Webpack Plugins
const ExtractTextPlugin = require('extract-text-webpack-plugin');

const extractCSS = new ExtractTextPlugin({filename: 'client/dist/styles/hail-carousel.bundle.css'});

module.exports = {
    entry: [
        './client/src/js/index.js',
    ],
    output: {
        filename: 'client/dist/js/hail-carousel.bundle.js',
        path: path.resolve(__dirname, './')
    },
    module: {
        rules: [
            {
                test: /\.js$/,
                exclude: /node_modules/,
                use: [
                    {
                        loader: 'babel-loader',
                        options: {
                            presets: [
                                "env",
                                "stage-2"
                            ]
                        }
                    }
                ]
            },
            {
                test: /\.scss$/,
                exclude: /node_modules/,
                use: extractCSS.extract({
                    use: [
                        {
                            loader: 'css-loader',
                            options: {
                                sourceMap: true
                            }
                        },
                        {
                            loader: 'postcss-loader',
                            options: {
                                sourceMap: true,
                                plugins: [
                                    autoprefixer
                                ]
                            }
                        },
                        {
                            loader: 'sass-loader',
                            options: {
                                sourceMap: true
                            }
                        }
                    ]
                })
            }, {
                test: /\.css$/,
                exclude: /node_modules/,
                use: extractCSS.extract({
                    use: [
                        {
                            loader: 'css-loader',
                            options: {
                                sourceMap: true
                            }
                        },
                        {
                            loader: 'postcss-loader',
                            options: {
                                sourceMap: true,
                                plugins: [
                                    autoprefixer
                                ]
                            }
                        }
                    ]
                })
            }, {
                test: /\.(png|ico|svg|jpg|gif)$/,
                exclude: /node_modules/,
                use: [
                    {
                        loader: 'file-loader',
                        options: {
                            name: '[name].[ext]',
                            outputPath: './client/dist/',
                            publicPath: "../images",
                            useRelativePath: true
                        }
                    }
                ]
            }, {
                test: /\.(ttf|eot|woff|woff2)$/,
                exclude: /node_modules/,
                use: [
                    {
                        loader: 'file-loader',
                        options: {
                            name: '[name].[ext]',
                            outputPath: './client/dist/',
                            publicPath: "../fonts",
                            useRelativePath: true
                        }
                    }
                ]
            }
        ]
    },
    plugins: [
        extractCSS
    ]
};


client/src/js/hailcarousel.js
jQuery(($) => {
    $('.hail-carousel').each((index, element) => {
        //Build carousel options (from the Hail carousel DataObject)
        var slickOptions = {
            lazyLoad: 'ondemand',
            dots: ($(element).data('showdots') === "Yes"),
            slidesToShow: $(element).data('lg-toshow'),
            slidesToScroll: $(element).data('lg-toscroll'),
            responsive: [
                {
                    breakpoint: $(element).data('md-break'),
                    settings: {
                        slidesToShow: $(element).data('md-toshow'),
                        slidesToScroll: $(element).data('md-toscroll')
                    }
                },
                {
                    breakpoint: $(element).data('sm-break'),
                    settings: {
                        slidesToShow: $(element).data('sm-toshow'),
                        slidesToScroll: $(element).data('sm-toscroll'),
                        arrows: false
                    }
                }
            ]
        };
        //Remove arrows padding in the wrapper for the lowest breakpoint
        var css = "@media (max-width:" + $(element).data('sm-break') + "px) { .hail-carousel-wrapper { padding-left: 0; padding-right: 0; } }",
            head = document.head || document.getElementsByTagName('head')[0],
            style = document.createElement('style');

        style.type = 'text/css';
        if (style.styleSheet) {
            style.styleSheet.cssText = css;
        } else {
            style.appendChild(document.createTextNode(css));
        }

        head.appendChild(style);

        //Create Slick carousel
        $(element).slick(slickOptions);
    });
});


src/Elements/HailListElement.php
<?php

namespace Firebrand\HailElemental\Elements;

use DNADesign\Elemental\Models\BaseElement;
use Firebrand\Hail\Pages\HailPage;
use Sheadawson\DependentDropdown\Forms\DependentDropdownField;
use SilverStripe\Forms\DropdownField;
use SilverStripe\Forms\NumericField;
use SilverStripe\Forms\RequiredFields;
use SilverStripe\Forms\Tab;
use SilverStripe\Forms\TabSet;
use SilverStripe\ORM\DataObject;
use SilverStripe\View\Requirements;

class HailCarouselElement extends BaseElement
{
    private static $table_name = 'HailCarouselElement';

    private static $db = [
        'FilterTags' => 'Text',
        'Limit' => 'Int',
        'ShowDots' => 'Enum(array("Yes","No"))',
        'LargeToShow' => 'Int',
        'LargeToScroll' => 'Int',
        'MediumBreakpoint' => 'Int',
        'MediumToShow' => 'Int',
        'MediumToScroll' => 'Int',
        'SmallBreakpoint' => 'Int',
        'SmallToShow' => 'Int',
        'SmallToScroll' => 'Int',
    ];

    private static $defaults = [
        'Limit' => 9,
        'ShowDots' => 'No',
        'LargeToShow' => 3,
        'LargeToScroll' => 3,
        'MediumBreakpoint' => 1024,
        'MediumToShow' => 2,
        'MediumToScroll' => 2,
        'SmallBreakpoint' => 600,
        'SmallToShow' => 1,
        'SmallToScroll' => 1,
    ];

    private static $singular_name = 'hail carousel';

    private static $plural_name = 'hail carousels';

    private static $description = 'Add a Hail Carousel';

    private static $has_one = [
        "HailPage" => "Firebrand\Hail\Pages\HailPage",
    ];

    public function getCMSFields()
    {
        $fields = parent::getCMSFields();

        //Multi tabs display
        $fields->addFieldToTab('Root.Main', TabSet::create('SubTabs', 'SubTabs', [Tab::create('Main'), Tab::create('Style')]));

        //Create new fields
        $limit = NumericField::create('Limit', 'Limit');
        $showdots = DropdownField::create('ShowDots', 'Show dots under the carousel', ['Yes' => 'Yes', 'No' => 'No']);
        $page = DropdownField::create('HailPageID', 'Hail Page', HailPage::get());
        $tag_source = function ($val) {
            $hailpage = DataObject::get_by_id(HailPage::class, $val);
            if ($hailpage) {
                return $hailpage->getAllowedPublicTags();
            }
            return [];
        };
        $filter_tags = DependentDropdownField::create('FilterTags', 'Public Tag displayed in the carousel', $tag_source)
            ->setDepends($page);

        $large_to_show = NumericField::create('LargeToShow', 'Large screens: Slides to show');
        $large_to_scroll = NumericField::create('LargeToScroll', 'Large screens: Slides to scroll');
        $medium_breakpoint = NumericField::create('MediumBreakpoint',
            'Medium screens: Breakpoint')->setDescription('The carousel will use the "Medium screens" settings when its width is under this value . In pixels . ');
        $medium_to_show = NumericField::create('MediumToShow', 'Medium screens: Slides to show');
        $medium_to_scroll = NumericField::create('MediumToScroll', 'Medium screens: Slides to scroll');
        $small_breakpoint = NumericField::create('SmallBreakpoint',
            'Small screens: Breakpoint')->setDescription('The carousel will use the "Small screens" settings when its width is under this value . In pixels . ');
        $small_to_show = NumericField::create('SmallToShow', 'Small screens: Slides to show');
        $small_to_scroll = NumericField::create('SmallToScroll', 'Small screens: Slides to scroll');

        //Add to tabs
        $fields->addFieldsToTab('Root.Main.SubTabs.Main', [
            $page,
            $filter_tags,
            $limit,
            $showdots,
        ]);
        $fields->addFieldsToTab('Root.Main.SubTabs.Style', [
            $large_to_show,
            $large_to_scroll,
            $medium_breakpoint,
            $medium_to_show,
            $medium_to_scroll,
            $small_breakpoint,
            $small_to_show,
            $small_to_scroll,
        ]);
        return $fields;
    }

    public function forTemplate($holder = true)
    {
        Requirements::css('resources/vendor/firebrand/silverstripe-hail-elemental/client/dist/styles/hail-carousel.bundle.css');
        Requirements::javascript('resources/vendor/firebrand/silverstripe-hail-elemental/client/dist/js/hail-carousel.bundle.js');

        return parent::forTemplate($holder);
    }

    public function getCMSValidator()
    {
        return new RequiredFields([
            'HailPageID',
        ]);
    }


    public function getType()
    {
        return 'Hail Carousel';
    }

    public function getCarouselItems()
    {
        $filter_tags = empty($this->FilterTags) || $this->FilterTags === "*" ? null : $this->FilterTags;

        return $this->HailPage()->getFullHailList($this->Limit, $filter_tags);
    }
}

src/Elements/HailCarouselElement.php
<?php

namespace Firebrand\HailElemental\Elements;

use DNADesign\Elemental\Models\BaseElement;
use Firebrand\Hail\Pages\HailPage;
use Sheadawson\DependentDropdown\Forms\DependentDropdownField;
use SilverStripe\Forms\DropdownField;
use SilverStripe\Forms\NumericField;
use SilverStripe\Forms\RequiredFields;
use SilverStripe\Forms\Tab;
use SilverStripe\Forms\TabSet;
use SilverStripe\ORM\DataObject;
use SilverStripe\ORM\PaginatedList;
use SilverStripe\View\Requirements;

/**
 * Hail Carousel Elemental element
 *
 * @package silverstripe-hail-elemental
 * @author Marc Espiard, Firebrand
 * @version 1.0
 *
 * @property string $FilterTags
 * @property int $Limit
 * @property string $ShowDots
 * @property int $LargeToShow
 * @property int $LargeToScroll
 * @property int $MediumBreakpoint
 * @property int $MediumToShow
 * @property int $MediumToScroll
 * @property int $SmallBreakpoint
 * @property int $SmallToShow
 * @property int $SmallToScroll
 */
class HailCarouselElement extends BaseElement
{
    private static $table_name = 'HailCarouselElement';
    private static $db = [
        'FilterTags' => 'Text',
        'Limit' => 'Int',
        'ShowDots' => 'Enum(array("Yes","No"))',
        'LargeToShow' => 'Int',
        'LargeToScroll' => 'Int',
        'MediumBreakpoint' => 'Int',
        'MediumToShow' => 'Int',
        'MediumToScroll' => 'Int',
        'SmallBreakpoint' => 'Int',
        'SmallToShow' => 'Int',
        'SmallToScroll' => 'Int',
    ];
    private static $defaults = [
        'Limit' => 9,
        'ShowDots' => 'No',
        'LargeToShow' => 3,
        'LargeToScroll' => 3,
        'MediumBreakpoint' => 1024,
        'MediumToShow' => 2,
        'MediumToScroll' => 2,
        'SmallBreakpoint' => 600,
        'SmallToShow' => 1,
        'SmallToScroll' => 1,
    ];
    private static $singular_name = 'hail carousel';
    private static $plural_name = 'hail carousels';
    private static $description = 'Add a Hail Carousel';
    private static $has_one = [
        "HailPage" => "Firebrand\Hail\Pages\HailPage",
    ];
    private static $icon = 'hail-icon';

    public function getCMSFields()
    {
        $fields = parent::getCMSFields();

        //Multi tabs display
        $fields->addFieldToTab('Root.Main', TabSet::create('SubTabs', 'SubTabs', [Tab::create('Main'), Tab::create('Style')]));

        //Create new fields
        $limit = NumericField::create('Limit', 'Limit');
        $showdots = DropdownField::create('ShowDots', 'Show dots under the carousel', ['Yes' => 'Yes', 'No' => 'No']);
        $page = DropdownField::create('HailPageID', 'Hail Page', HailPage::get());
        $tag_source = function ($val) {
            $hailpage = DataObject::get_by_id(HailPage::class, $val);
            if ($hailpage) {
                return $hailpage->getAllowedPublicTags();
            }
            return [];
        };
        $filter_tags = DependentDropdownField::create('FilterTags', 'Public Tag displayed in the carousel', $tag_source)
            ->setDepends($page);

        $large_to_show = NumericField::create('LargeToShow', 'Large screens: Slides to show');
        $large_to_scroll = NumericField::create('LargeToScroll', 'Large screens: Slides to scroll');
        $medium_breakpoint = NumericField::create('MediumBreakpoint',
            'Medium screens: Breakpoint')->setDescription('The carousel will use the "Medium screens" settings when its width is under this value . In pixels . ');
        $medium_to_show = NumericField::create('MediumToShow', 'Medium screens: Slides to show');
        $medium_to_scroll = NumericField::create('MediumToScroll', 'Medium screens: Slides to scroll');
        $small_breakpoint = NumericField::create('SmallBreakpoint',
            'Small screens: Breakpoint')->setDescription('The carousel will use the "Small screens" settings when its width is under this value . In pixels . ');
        $small_to_show = NumericField::create('SmallToShow', 'Small screens: Slides to show');
        $small_to_scroll = NumericField::create('SmallToScroll', 'Small screens: Slides to scroll');

        //Add to tabs
        $fields->addFieldsToTab('Root.Main.SubTabs.Main', [
            $page,
            $filter_tags,
            $limit,
            $showdots,
        ]);
        $fields->addFieldsToTab('Root.Main.SubTabs.Style', [
            $large_to_show,
            $large_to_scroll,
            $medium_breakpoint,
            $medium_to_show,
            $medium_to_scroll,
            $small_breakpoint,
            $small_to_show,
            $small_to_scroll,
        ]);
        return $fields;
    }

    public function forTemplate($holder = true)
    {
        Requirements::css('firebrandhq/silverstripe-hail-elemental: client/dist/styles/hail-carousel.bundle.css');

        //Require jquery and bootstrap from Hail module to avoid double inclusion and to be able to globally disable this requirements for all hail modules
        Requirements::css('firebrandhq/silverstripe-hail: thirdparty/bootstrap/styles/bootstrap.min.css');
        Requirements::javascript('firebrandhq/silverstripe-hail: thirdparty/jquery/js/jquery.min.js');
        Requirements::javascript('firebrandhq/silverstripe-hail: thirdparty/bootstrap/js/bootstrap.bundle.min.js');

        Requirements::javascript('firebrandhq/silverstripe-hail-elemental: client/dist/js/hail-carousel.bundle.js');

        return parent::forTemplate($holder);
    }

    public function getCMSValidator()
    {
        return new RequiredFields([
            'HailPageID',
        ]);
    }


    public function getType()
    {
        return 'Hail Carousel';
    }

    /**
     * Get the carousel items filtering them according to the element configuration
     *
     * @return PaginatedList
     */
    public function getCarouselItems()
    {
        $filter_tags = empty($this->FilterTags) || $this->FilterTags === "*" ? null : $this->FilterTags;

        return $this->HailPage()->getFullHailList($this->Limit, $filter_tags);
    }
}